### Cleaning the Music Genre Dataset

In [11]:
import pandas as pd
import numpy as np
import re

#lendo o arquivo allgenres
genres = pd.read_excel(r'C:\Users\Nathan\Documents\music recommendation system\allgenres.xlsx',header=None)
genres = genres[1]

In [12]:
genres

0                          rumba,flamenco guitar,flamenco
1                                         rap,chicago rap
2                                              neo mellow
3       permanent wave,rock,post-grunge,alternative ro...
4       rock en espanol,musica aragonesa,latin rock,la...
                              ...                        
3370    chicago rap,alternative hip hop,escape room,so...
3371                                                  NaN
3372    christian music,worship,deep christian rock,ro...
3373      harlem hip hop,hardcore hip hop,hip pop,hip hop
3374    glam rock,pub rock,art rock,classic uk pop,sym...
Name: 1, Length: 3375, dtype: object

In [13]:
all_genres = []

for i in range(0,len(genres)):
    if type(genres[i]) != float:
        if type(re.search(',',genres[i])) == re.Match:
            str1 = re.findall('(.*?),',genres[i])
            str2 = re.findall(',([^,]*)$',genres[i])

            for j in range(0,len(str1)):
                all_genres.append(str1[j])

            all_genres.append(str2[0])

        else:
            all_genres.append(genres[i])
            

In [14]:
unique_genres = pd.Series(all_genres)
unique_genres = unique_genres.unique()
unique_genres = pd.Series(unique_genres)

In [15]:
print(len(unique_genres),len(all_genres))

1541 12997


In [16]:
all_genres = pd.Series(all_genres)

In [17]:
top_genres = all_genres.value_counts()

In [18]:
top_genres.head()

rock                 291
pop rock             191
dance pop            191
classic rock         143
alternative metal    142
dtype: int64

For there is too many genres in this dataset (1541 to be exact), it might be good to reduce some of this genres for another that is more generic. For example, reducing "Classic Rock" to "Rock" or "Indie Pop" to "Indie" 

   
For that we will be doing substitue_genres function


In [9]:
def substitute_genres(string_list):
    for j in range(0,len(string_list)):
                if type(re.search('rap',string_list[j])) == re.Match:
                    string_list[j] = 'rap'
                if type(re.search('hip-hop',string_list[j])) == re.Match:
                    string_list[j] = 'rap'
                if type(re.search('rock',string_list[j])) == re.Match:
                    string_list[j] = 'rock'
                if type(re.search('indie',string_list[j])) == re.Match:
                    string_list[j] = 'indie'
                if type(re.search('hip hop',string_list[j])) == re.Match:
                    string_list[j] = 'rap'
                if type(re.search('pop',string_list[j])) == re.Match:
                    string_list[j] = 'pop'
                if type(re.search('metal',string_list[j])) == re.Match:
                    string_list[j] = 'metal'
                if type(re.search('dance',string_list[j])) == re.Match:
                    string_list[j] = 'pop'
                if type(re.search('country',string_list[j])) == re.Match:
                    string_list[j] = 'country'
                if type(re.search('blues',string_list[j])) == re.Match:
                    string_list[j] = 'blues'
                if type(re.search('folk',string_list[j])) == re.Match:
                    string_list[j] = 'folk'
                if type(re.search('metal',string_list[j])) == re.Match:
                    string_list[j] = 'metal'
                if type(re.search('punk',string_list[j])) == re.Match:
                    string_list[j] = 'punk'
                if type(re.search('eletronica',string_list[j])) == re.Match:
                    string_list[j] = 'eletronic'
                if type(re.search('eletronic',string_list[j])) == re.Match:
                    string_list[j] = 'eletronic'
                if type(re.search('jazz',string_list[j])) == re.Match:
                    string_list[j] = 'jazz'
                if type(re.search('soul',string_list[j])) == re.Match:
                    string_list[j] = 'soul'
                if type(re.search('funk',string_list[j])) == re.Match:
                    string_list[j] = 'reggae'
                if type(re.search('screamo',string_list[j])) == re.Match:
                    string_list[j] = 'hardcore'
                if type(re.search('hardcore',string_list[j])) == re.Match:
                    string_list[j] = 'hardcore'
                if type(re.search('trap',string_list[j])) == re.Match:
                    string_list[j] = 'trap'
                if type(re.search('disco',string_list[j])) == re.Match:
                    string_list[j] = 'disco'
                if type(re.search('edm',string_list[j])) == re.Match:
                    string_list[j] = 'eletronic'
                if type(re.search('classical',string_list[j])) == re.Match:
                    string_list[j] = 'classical'
                if type(re.search('r&b',string_list[j])) == re.Match:
                    string_list[j] = 'r&b'
                if type(re.search('rhythm and blues',string_list[j])) == re.Match:
                    string_list[j] = 'r&b'
                if type(re.search('grunge',string_list[j])) == re.Match:
                    string_list[j] = 'grunge'
                if type(re.search('swing',string_list[j])) == re.Match:
                    string_list[j] = 'jazz'
                if type(re.search('emo',string_list[j])) == re.Match:
                    string_list[j] = 'hardcore'
                if type(re.search('gospel',string_list[j])) == re.Match:
                    string_list[j] = 'gospel'
                if type(re.search('christian',string_list[j])) == re.Match:
                    string_list[j] = 'gospel'
                if type(re.search('samba',string_list[j])) == re.Match:
                    string_list[j] = 'samba'
                if type(re.search('bossa nova',string_list[j])) == re.Match:
                    string_list[j] = 'bossa nova'
                if type(re.search('worship',string_list[j])) == re.Match:
                    string_list[j] = 'gospel'
                if type(re.search('baroque',string_list[j])) == re.Match:
                    string_list[j] = 'classical'
                if type(re.search('trance',string_list[j])) == re.Match:
                    string_list[j] = 'electronic'
                if type(re.search('deathrash',string_list[j])) == re.Match:
                    string_list[j] = 'metal'
                if type(re.search('ebm',string_list[j])) == re.Match:
                    string_list[j] = 'electronic'
                if type(re.search('doom',string_list[j])) == re.Match:
                    string_list[j] = 'metal'
                if type(re.search('techno',string_list[j])) == re.Match:
                    string_list[j] = 'electronic'
                if type(re.search('flamenco',string_list[j])) == re.Match:
                    string_list[j] = 'flamenco' 
                if type(re.search('latin',string_list[j])) == re.Match:
                    string_list[j] = 'latin'
                if type(re.search('house',string_list[j])) == re.Match:
                    string_list[j] = 'electronic'
                if type(re.search('eletronic',string_list[j])) == re.Match:
                    string_list[j] = 'electronic'    
                if type(re.search('reggae',string_list[j])) == re.Match:
                    string_list[j] = 'reggae'

    return string_list

In [10]:
all_genres = substitute_genres(all_genres)

### Creating a list of lists to group the genres and take off the repeated ones 

In [13]:
for i in range(0,len(genres)):
    if type(genres[i]) != float:
        if type(re.search(',',genres[i])) == re.Match:
            str1 = re.findall('(.*?),',genres[i])
            str1 = substitute_genres(str1)  
                
            str2 = re.findall(',([^,]*)$',genres[i])
            str2 = substitute_genres(str2)
            
            
            genres[i] = str1 + str2
            genres[i] = list(dict.fromkeys(genres[i]))
        else:
            genres_aux = []
            genres_aux.append(genres[i])
            genres[i] = genres_aux
            genres[i] = substitute_genres(genres[i])
            genres[i] = list(dict.fromkeys(genres[i]))
            

In [15]:
# clean genres excel
genres.to_excel(r'C:\Users\Nathan\Documents\music recommendation system\clean_genres.xlsx',header=None)

### Cleaning the genres from list to string

In [9]:
import pandas as pd
import numpy as np

genres = pd.read_excel(r'C:\Users\Nathan\Documents\music recommendation system\clean_genres.xlsx',header=None)
genres = genres[1]

In [10]:
len(genres)

3375

In [6]:
import re
str1 = re.findall("'(.*?)'",genres[0])

In [7]:
for i in range(0,len(genres)):
    if type(genres[i]) != float:
        genre = ''
        str1 = re.findall("'(.*?)'",genres[i])
        for j in range(0,len(str1)):
            genre = genre + '|' + str1[j]
        genre = genre[1:]
        genres[i] = genre

In [8]:
genres

0                         rumba|flamenco
1                                    rap
2                             neo mellow
3       permanent wave|rock|grunge|metal
4            rock|musica aragonesa|latin
                      ...               
3370                     rap|escape room
3371                                 NaN
3372      gospel|rock|ccm|classic praise
3373                             rap|pop
3374                            rock|pop
Name: 1, Length: 3375, dtype: object

In [69]:
genres.to_excel(r'C:\Users\Nathan\Documents\music recommendation system\clean_genres2.xlsx',header=None)

In [2]:
import pandas as pd
import numpy as np

allgenres = pd.read_excel(r'C:\Users\Nathan\Documents\music recommendation system\clean_genres2.xlsx',header=None)

In [4]:
allgenres.tail(20)

0                                                1
3355  3355  rock|pop|mellow gold|disco|adult standards|folk
3356  3356                                             folk
3357  3357                                              NaN
3358  3358                                         jam band
3359  3359                                 rock|indie|blues
3360  3360                               pop|latin|tropical
3361  3361                          pop|latin|rock|tropical
3362  3362                    swedish electronic|electronic
3363  3363                                             rock
3364  3364                                              NaN
3365  3365                                      soul|motown
3366  3366                                     new rave|pop
3367  3367                                         hands up
3368  3368                                   rock|pop|indie
3369  3369                                              rap
3370  3370                                  rap|escape room
3371  3371                                              NaN
3372  3372                   gospel|rock|ccm|classic praise
3373  3373                                          rap|pop
3374  3374                                         rock|pop